### ISSUE TITLE
What's the difference between 'val' mode and 'test' mode in matconvnet?

### ISSUE THREAD

**Issue by MichaelLiang12** [vlfeat/matconvnet#538]
Is the 'val' mode the general meaning (cross-validation, use a  set of examples to tune the parameters) in matconvnet codes ? Or it is the same as 'test' mode (use a set of examples  only to assess the performance)?

For example, codes below is part of cnn_train.m, why should we choose the 'test' mode if it is not 'train' mode, is there no 'val' mode?
```` matlab
% part of cnn_train.m
 if strcmp(mode, 'train')
      dzdy = 1 ;
      evalMode = 'normal' ;
else
      dzdy = [] ;
      evalMode = 'test' ;
end
```
Another example it's the codes in cnn_mnist.m. If imdb.images.set == 3, then those images belong to test sets, why give them a label of 'val' ?
``` matlab
[net, info] = trainfn(net, imdb, getBatch(opts), ...
  'expDir', opts.expDir, ...
  net.meta.trainOpts, ...
  opts.train, ...
  'val', find(imdb.images.set == 3)) ;
```

**Comment by fengyunxiaozi** [comment]
I was also confused about this. a possible explanation might be: matconvnet cost large memory during training, so it cannot work efficiently as caffe. it seems matconvnet default there are only train and test data, and this can be seen when preparing imdb.   @vedaldi @lenck  is there more reasonable explantion? thanks!

**Comment by vedaldi** [comment]
There is a difference between train, val, and test sets, which are just data splits, and running a network in 'normal' and 'test' mode. You would run on 'test' mode on both the val and test data spolits. The only difference between running in 'normal' and 'test' mode is that certain layers behave differently (e.g. dropout is skipped, batch normalisation uses the average moments instead of computing them on the fly). This is done because this is how the network would behave once deployed.

In term of efficiency, in general we noticed that MCN tends to use less memory than Caffe because we delete buffers aggressively (this is particularly true for fully connected networks), but that usually has a small cost in term of running time.
